tools/pepsi/src/upload.rs

```diff
@@ -63,58 +121,48 @@ pub async fn upload(arguments: Arguments, repository: &Repository) -> Result<()>
         .wrap_err("failed to build the code")?;
     }
 
-    communication(
-        match arguments.no_communication {
-            true => CommunicationArguments::Disable { nao_numbers },
-            false => CommunicationArguments::Enable { nao_numbers },
-        },
-        repository,
-    )
-    .await
-    .wrap_err("failed to set communication enablement directory")?;
-
     let (_temporary_directory, hulk_directory) = repository
-        .create_upload_directory(arguments.profile)
+        .create_upload_directory(arguments.profile.as_str())
         .await
         .wrap_err("failed to create upload directory")?;
 
-    if !arguments.no_restart {
-        println!("Stopping HULK");
-        hulk(HulkArguments {
-            action: SystemctlAction::Stop,
-            naos: arguments.naos.clone(),
-        })
+    let hardware_ids = repository
+        .get_hardware_ids()
         .await
-        .wrap_err("failed to stop HULK service")?;
-    }
+        .wrap_err("failed to get hardware IDs")?;
 
-    let tasks = arguments.naos.iter().map(|nao_address| {
-        let hulk_directory = hulk_directory.clone();
-        async move {
-            let nao = Nao::new(nao_address.ip);
+    let multi_progress = ProgressIndicator::new();
 
-            if !arguments.skip_os_check && !nao.has_stable_os_version().await {
-                return Ok(());
-            }
-
-            println!("Starting upload to {nao_address}");
-            nao.upload(hulk_directory, !arguments.no_clean)
-                .await
-                .wrap_err_with(|| format!("failed to power {nao_address} off"))
-        }
-    });
-
-    let results = join_all(tasks).await;
-    gather_results(results, "failed to execute some upload tasks")?;
+    arguments
+        .naos
+        .iter()
```

Comment:
Why does this not use your new infrastructure?

Reply:
This is due to `upload_with_progress` requiring a reference to `progress`. I was unable to rewrite `map_tasks()` in such a way to accommodate that due to the closure being executed in an async context.